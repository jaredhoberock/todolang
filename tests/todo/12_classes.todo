{
  class Breakfast {
    fun cook() {
      print "Eggs a-fryin'!";
    }
  
    fun serve(who) {
      print "Enjoy your breakfast, " + who + ".";
    }
  }
}

{
  class DevonshireCream {
    fun serveOn() {
      return "Scones";
    }
  }

  print DevonshireCream;
}

{
  class Bagel {}
  var bagel = Bagel();
  print bagel;
}

{
  class Box {}

  fun notMethod(argument) {
    print "called function with " + argument;
  }

  var box = Box();
  box.function = notMethod;
  box.function("argument");
}

{
  class Bacon {
    fun eat() {
      print "Crunch crunch crunch!";
    }
  }

  Bacon().eat();
}

{
  class Egotist {
    fun speak() {
      print this;
    }
  }

  var method = Egotist().speak;
  method();
}

{
  class Cake {
    fun taste() {
      var adjective = "delicious";
      print "The " + this.flavor + " cake is " + adjective + "!";
    }
  }

  var cake = Cake();
  cake.flavor = "German chocolate";
  cake.taste(); // Prints "The German chocolate cake is delicious!".
}

{
  class Thing {
    fun getCallback() {
      fun localFunction() {
        print this;
      }

      return localFunction;
    }
  }

  var callback = Thing().getCallback();
  callback();
}

{
  class Person {
    fun sayName() {
      print this.name;
    }
  }
  
  var jane = Person();
  jane.name = "Jane";
  
  var method = jane.sayName;
  method();
}

{
  class Person {
    fun sayName() {
      print this.name;
    }
  }

  var jane = Person();
  jane.name = "Jane";

  var bill = Person();
  bill.name = "Bill";

  bill.sayName = jane.sayName;
  bill.sayName();
}

{
  class Foo {
    fun init() {
      print this;
    }
  }

  var foo = Foo();
  print foo.init();
}

{
  class Bar {
    fun init() {
      return;
    }
  }

  var bar = Bar();
  print bar.init();
}

print "OK";
