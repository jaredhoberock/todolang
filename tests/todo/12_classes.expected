<class DevonshireCream>
<class Bagel> instance
called function with argument
Crunch crunch crunch!
<class Egotist> instance
The German chocolate cake is delicious!
<class Thing> instance
Jane
Jane
<class Foo> instance
<class Foo> instance
<class Foo> instance
<class Bar> instance
OK
